kraken-std/src/kraken/std/buffrs/tasks.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+import contextlib
+import enum
+import logging
+import os
+import subprocess as sp
+
+from kraken.common import CredentialsWithHost, atomic_file_swap
+from kraken.core import BackgroundTask, Project, Property, Task, TaskStatus
+
+from .manifest import BuffrsManifest
+
+logger = logging.getLogger(__name__)
+
+
+class Language(enum.Enum):
+    PYTHON = "python"
+
+
+class BuffrsLoginTask(Task):
+    """This task logs into artifactory with buffrs"""
+
+    description = "Login to artifactory with buffrs."
+    artifactory_credentials: Property[CredentialsWithHost]
+
+    def execute(self) -> TaskStatus:
+        credentials = self.artifactory_credentials.get()
+
+        command = ["buffrs", "login", "--registry", credentials.host]
+
+        return TaskStatus.from_exit_code(
+            command,
+            sp.run(
+                command,
+                cwd=self.project.directory,
+                env=os.environ.copy(),
+                input=credentials.password,
+                text=True,
+            ).returncode,
+        )
+
+
+class BuffrsInstallTask(Task):
+    """Install dependencies defined in `Proto.toml`"""
+
+    description = "Runs `buffrs install` to download protobuf dependencies"
+
+    def execute(self) -> TaskStatus:
+        command = ["buffrs", "install"]
+
+        return TaskStatus.from_exit_code(
+            command,
+            sp.call(
+                command,
+                cwd=self.project.directory,
+                env=os.environ.copy(),
+            ),
+        )
+
+
+class BuffrsPublishTask(BackgroundTask):
```

Comment:
This doesn't need to be a background task now. You just need to move the `atomic_file_swap()` call into the `execute()` method and wrap it around the Buffrs command invokation.